### PR TITLE
UCP/MM: Selectively exclude mem type MDs from rcache

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**
## What
Bypass UCP rcache when using gdr_copy for mem_type operations.

## Why?
gdr_copy has its own rcache and double rcache restores internal issue [3606732](https://redmine.mellanox.com/issues/3606732). The root cause of this issue is the insufficient support for overlapping regions in the gdr_copy driver. With single caching, the memory handle's reference count tracks the number of in-flight operations, ensuring no overlapping memory regions for mem_type operations. However, with double caching, the reference count in the secondary cache corresponds to the number of primary memory regions, which does not provide the same guarantee.

## How?
Introduce a static user memh flow for mem_type operations and mark gdr_copy as non-cacheable of UCP.

### CODE REVIEW

File: src/ucp/core/ucp_mm.h [hunk 1]
```diff
@@ -27,6 +27,8 @@
    existing region */
 #define UCP_MM_UCT_ACCESS_MASK UCT_MD_MEM_ACCESS_ALL
 
+#define UCP_MM_UCT_NON_CACHED UCS_BIT(31)
```
Comment (brminich) [inline src/ucp/core/ucp_mm.h:30]: imo, such flag needs to be defined in UCT instead

File: src/uct/cuda/gdr_copy/gdr_copy_md.c [hunk 2]
```diff
@@ -62,7 +62,7 @@ uct_gdr_copy_md_query(uct_md_h uct_md, uct_md_attr_v2_t *md_attr)
     uct_md_base_md_query(md_attr);
     md_attr->flags            = UCT_MD_FLAG_REG | UCT_MD_FLAG_NEED_RKEY;
     md_attr->reg_mem_types    = UCS_BIT(UCS_MEMORY_TYPE_CUDA);
-    md_attr->cache_mem_types  = UCS_BIT(UCS_MEMORY_TYPE_CUDA);
+    md_attr->cache_mem_types  = 0;
```
Comment (brminich) [inline src/uct/cuda/gdr_copy/gdr_copy_md.c:65]: should not be this enough to avoid UCP rcache usage?

File: src/ucp/core/ucp_mm.c [hunk 3]
```diff
@@ -825,6 +826,7 @@ static ucs_status_t ucp_memh_init_uct_reg(ucp_context_h context, ucp_mem_h memh,
         }
 
         ucp_memh_init_from_parent(memh, cache_md_map);
+        memh->flags |= UCP_MEMH_FLAG_NON_CACHED;
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:829]: we set this in both if/else, should it be outside to do it only once?
or even do it from ucp_mem_map/ucp_memh_create_from_mem/ucp_memh_import when they call ucp_memh_create?

File: src/ucp/core/ucp_mm.c [hunk 4]
```diff
@@ -1140,7 +1143,7 @@ ucs_status_t ucp_mem_map(ucp_context_h context, const ucp_mem_map_params_t *para
 
 void ucp_memh_put_slow(ucp_context_h context, ucp_mem_h memh)
 {
-    ucs_assert(context->rcache == NULL);
+    ucs_assert(memh->flags & UCP_MEMH_FLAG_NON_CACHED);
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:1146]: shall we do also an (inverse) assert when releasing memh from rcache?

File: src/ucp/core/ucp_mm.c [hunk 5]
```diff
@@ -1180,9 +1184,12 @@ ucs_status_t ucp_mem_type_reg_buffers(ucp_worker_h worker, void *remote_addr,
     tl_md  = &context->tl_mds[md_index];
     cmpt   = context->tl_cmpts[tl_md->cmpt_index].cmpt;
 
+    if (!(context->cache_md_map[mem_type] & UCS_BIT(md_index))) {
+        uct_flags |= UCP_MM_UCT_NON_CACHED;
+    }
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:1189]: can we combine cached and non-cached MDs in the same memh?

File: src/ucp/core/ucp_mm.h (hunk 1 above)
Comment (yosefe) [reply src/ucp/core/ucp_mm.h:30]: can we avoid this flag, since we already have UCP_MEMH_FLAG_NON_CACHED?

File: src/ucp/dt/dt.c [hunk 6]
```diff
@@ -31,10 +31,10 @@ UCS_PROFILE_FUNC_VOID(ucp_mem_type_unpack,
                       ucp_worker_h worker, void *buffer, const void *recv_data,
                       size_t recv_length, ucs_memory_type_t mem_type)
 {
-    ucp_ep_h ep         = worker->mem_type_ep[mem_type];
+    ucp_ep_h ep                 = worker->mem_type_ep[mem_type];
+    ucp_mem_dummy_handle_t memh = {};
```
Comment (brminich) [inline src/ucp/dt/dt.c:35]: why is this needed?
(also in `ucp_mem_type_pack`)

Comment (Artemy-Mellanox) [reply src/ucp/dt/dt.c:35]: in PR desc I call it "static user memh"

File: src/ucp/core/ucp_mm.c [hunk 7]
```diff
@@ -843,6 +841,29 @@ static ucs_status_t ucp_memh_init_uct_reg(ucp_context_h context, ucp_mem_h memh,
...
+    ucp_md_map_t reg_md_map    = context->reg_md_map[mem_type];
+    ucs_status_t status;
+
+    if (uct_flags & UCT_MD_MEM_FLAG_LOCK) {
+        reg_md_map |= context->reg_block_md_map[mem_type];
+    }
+
+    reg_md_map &= ~memh->md_map;
+    status = ucp_memh_init_uct_reg(context, memh, reg_md_map, uct_flags,
+                                   alloc_name);
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:858]: maybe pass "reg_md_map &  ~memh->md_map" to ucp_memh_init_uct_reg instead of modifying reg_md_map?

File: src/ucp/core/ucp_mm.c [hunk 8]
```diff
@@ -843,6 +841,29 @@ static ucs_status_t ucp_memh_init_uct_reg(ucp_context_h context, ucp_mem_h memh,
...
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    ucs_assert(ucp_memh_is_user_memh(memh));
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:863]: maybe do it in the beginning of the function?

File: src/ucp/core/ucp_mm.c [hunk 9]
```diff
@@ -1177,12 +1197,15 @@ ucs_status_t ucp_mem_type_reg_buffers(ucp_worker_h worker, void *remote_addr,
         goto out;
     }
 
+    memh->super.super.start = (uintptr_t)remote_addr;
+    memh->super.super.end   = (uintptr_t)remote_addr + length;
+    ucp_memh_init(memh, context, 0, UCT_MD_MEM_ACCESS_ALL,
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:1202]: can we skip ucp_memory_detect that is inside ucp_memh_init?

File: src/ucp/dt/dt.c (hunk 6 above)
Comment (yosefe) [reply src/ucp/dt/dt.c:35]: why need to initialize it to 0?

File: src/ucp/dt/dt.c [hunk 10]
```diff
@@ -59,19 +59,19 @@ UCS_PROFILE_FUNC_VOID(ucp_mem_type_unpack,
...
 }
 
 UCS_PROFILE_FUNC_VOID(ucp_mem_type_pack,
                       (worker, dest, src, length, mem_type),
                       ucp_worker_h worker, void *dest, const void *src,
                       size_t length, ucs_memory_type_t mem_type)
 {
-    ucp_ep_h ep         = worker->mem_type_ep[mem_type];
+    ucp_ep_h ep                 = worker->mem_type_ep[mem_type];
+    ucp_mem_dummy_handle_t memh = {};
```
Comment (yosefe) [inline src/ucp/dt/dt.c:71]: why need to initialize?

File: src/uct/cuda/gdr_copy/gdr_copy_md.c (hunk 2 above)
Comment (yosefe) [reply src/uct/cuda/gdr_copy/gdr_copy_md.c:66]: shall we remove line 65?

File: src/ucp/dt/dt.c (hunk 6 above)
Comment (Artemy-Mellanox) [reply src/ucp/dt/dt.c:35]: user memh flow expect it to be zeroed

Comment (yosefe) [reply src/ucp/dt/dt.c:35]: can we avoid this extra overhead?

File: src/ucp/dt/dt.c [hunk 11]
```diff
@@ -59,30 +60,31 @@ UCS_PROFILE_FUNC_VOID(ucp_mem_type_unpack,
...
 }
 
 UCS_PROFILE_FUNC_VOID(ucp_mem_type_pack,
                       (worker, dest, src, length, mem_type),
                       ucp_worker_h worker, void *dest, const void *src,
                       size_t length, ucs_memory_type_t mem_type)
 {
-    ucp_ep_h ep                 = worker->mem_type_ep[mem_type];
-    ucp_mem_dummy_handle_t memh = {};
+    ucp_ep_h ep         = worker->mem_type_ep[mem_type];
```
Comment (yosefe) [inline src/ucp/dt/dt.c:71]: can remove some spaces before =

File: src/ucp/dt/dt.c [hunk 12]
```diff
@@ -31,22 +31,23 @@ UCS_PROFILE_FUNC_VOID(ucp_mem_type_unpack,
                       ucp_worker_h worker, void *buffer, const void *recv_data,
                       size_t recv_length, ucs_memory_type_t mem_type)
 {
-    ucp_ep_h ep                 = worker->mem_type_ep[mem_type];
-    ucp_mem_dummy_handle_t memh = {};
+    ucp_ep_h ep         = worker->mem_type_ep[mem_type];
```
Comment (yosefe) [inline src/ucp/dt/dt.c:34]: can remove some spaces before =

**Review by yosefe — APPROVED** [review]

File: src/ucp/core/ucp_mm.c [hunk 13]
```diff
@@ -417,9 +416,19 @@ static void ucp_memh_put_rcache(ucp_context_h context, ucp_mem_h memh)
     UCP_THREAD_CS_EXIT(&context->mt_lock);
 }
 
+static void ucp_memh_dereg_all(ucp_context_h context, ucp_mem_h memh)
+{
+    if (memh->parent == memh) {
+        ucp_memh_dereg(context, memh, memh->md_map);
+    } else {
```
Comment (iyastreb) [inline src/ucp/core/ucp_mm.c:423]: Should we handle here the case when parent == NULL?

**Review by yosefe — APPROVED** [review]
